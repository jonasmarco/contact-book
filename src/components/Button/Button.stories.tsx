import { Meta, StoryFn } from '@storybook/react'

import { HeartFill, Peace } from '@styled-icons/bootstrap'

import Button, { Props } from '.'

export default {
  title: 'Button',
  component: Button,
  args: {
    children: 'Click me!',
    size: 'normal',
    category: 'primary'
  },
  argTypes: {
    children: {
      type: 'string'
    },
    size: {
      options: ['normal', 'large'],
      control: { type: 'radio' }
    },
    category: {
      options: ['primary', 'secondary', 'danger', 'success', 'warning'],
      control: { type: 'radio' }
    },
    fullWidth: {
      type: 'boolean',
      defaultValue: false
    },
    disabled: {
      type: 'boolean',
      defaultValue: false
    },
    otherSide: {
      type: 'boolean',
      defaultValue: false
    },
    as: {
      table: {
        disable: true
      }
    }
  }
} as Meta<typeof Button>

const Template: StoryFn<typeof Button> = (args: Props) => <Button {...args} />

export const Primary = Template.bind({})
Primary.args = {
  children: 'Primary'
}

export const Secondary = Template.bind({})
Secondary.args = {
  category: 'secondary'
}

export const Danger = Template.bind({})
Danger.args = {
  children: "I'm Danger",
  category: 'danger'
}

export const Success = Template.bind({})
Success.args = {
  children: 'Success for us',
  category: 'success'
}

export const Warning = Template.bind({})
Warning.args = {
  children: 'Pay attention',
  category: 'warning'
}

export const Large = Template.bind({})
Large.args = {
  children: "I'm a little bigger",
  size: 'large'
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  children: 'I"m a full width button and I"ll fill the entire screen',
  fullWidth: true
}

export const Disabled = Template.bind({})
Disabled.args = {
  children: 'You can"t click on me',
  category: 'secondary',
  disabled: true
}

export const WithIcon = Template.bind({})
WithIcon.args = {
  children: 'Love',
  icon: <HeartFill />
}

export const WithTheIconOnTheLeft = Template.bind({})
WithTheIconOnTheLeft.args = {
  children: 'Good vibes only',
  category: 'secondary',
  icon: <Peace />,
  otherSide: true
}
