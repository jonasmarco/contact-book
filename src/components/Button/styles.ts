import styled, { css, DefaultTheme } from 'styled-components'
import { Props } from '.'
import { customMedia } from '../MediaMatch'

type WrapperProps = {
  hasIcon: boolean
} & Pick<Props, 'size' | 'category' | 'fullWidth' | 'otherSide'>

const wrapperModifiers = {
  normal: (theme: DefaultTheme) => css`
    font-size: ${theme.fonts.sizes.normal};
    min-height: 4rem;
  `,
  large: (theme: DefaultTheme) => css`
    font-size: ${theme.fonts.sizes.large};
    min-height: 5rem;
  `,

  primary: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
    border: 1px solid ${theme.colors.primary};
    color: ${theme.colors.white};
    position: relative;
    z-index: ${theme.layers.behindMenu};
    span {
      line-height: ${theme.text.lineHeight};
    }
    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: ${theme.colors.gray};
      border-color: ${theme.colors.gray};
      box-shadow: 0 15px 20px rgba(0, 0, 0, 0.16), 0 5px 8px rgba(0, 0, 0, 0.08);
      span {
        color: ${theme.colors.white};
      }
      svg {
        fill: ${theme.colors.white};
      }
    }
  `,
  secondary: (theme: DefaultTheme) => css`
    background: ${theme.colors.secondary};
    border: 1px solid ${theme.colors.secondary};
    color: ${theme.colors.white};
    position: relative;
    z-index: ${theme.layers.behindMenu};
    span {
      line-height: ${theme.text.lineHeight};
    }
    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: ${theme.colors.blue};
      border-color: ${theme.colors.blue};
      box-shadow: 0 15px 20px rgba(0, 0, 0, 0.16), 0 5px 8px rgba(0, 0, 0, 0.08);
      span {
        color: ${theme.colors.white};
      }
      svg {
        fill: ${theme.colors.white};
      }
    }
  `,
  danger: (theme: DefaultTheme) => css`
    background: ${theme.colors.red};
    border: 1px solid ${theme.colors.red};
    color: ${theme.colors.white};
    position: relative;
    z-index: ${theme.layers.behindMenu};
    span {
      line-height: ${theme.text.lineHeight};
    }
    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: ${theme.colors.black};
      border-color: ${theme.colors.black};
      box-shadow: 0 15px 20px rgba(0, 0, 0, 0.16), 0 5px 8px rgba(0, 0, 0, 0.08);
      span {
        color: ${theme.colors.white};
      }
      svg {
        fill: ${theme.colors.white};
      }
    }
  `,
  success: (theme: DefaultTheme) => css`
    background: ${theme.colors.green};
    border: 1px solid ${theme.colors.green};
    color: ${theme.colors.white};
    position: relative;
    z-index: ${theme.layers.behindMenu};
    span {
      line-height: ${theme.text.lineHeight};
    }
    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: ${theme.colors.black};
      border-color: ${theme.colors.black};
      box-shadow: 0 15px 20px rgba(0, 0, 0, 0.16), 0 5px 8px rgba(0, 0, 0, 0.08);
      span {
        color: ${theme.colors.white};
      }
      svg {
        fill: ${theme.colors.white};
      }
    }
  `,
  warning: (theme: DefaultTheme) => css`
    background: ${theme.colors.yellow};
    border: 1px solid ${theme.colors.yellow};
    color: ${theme.colors.white};
    position: relative;
    z-index: ${theme.layers.behindMenu};
    span {
      line-height: ${theme.text.lineHeight};
    }
    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: ${theme.colors.black};
      border-color: ${theme.colors.black};
      box-shadow: 0 15px 20px rgba(0, 0, 0, 0.16), 0 5px 8px rgba(0, 0, 0, 0.08);
      span {
        color: ${theme.colors.white};
      }
      svg {
        fill: ${theme.colors.white};
      }
    }
  `,

  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (otherSide: boolean) => css`
    align-items: center;
    justify-content: center;
    span {
      margin: ${otherSide === false ? '0 1rem 0 0' : '0 0 0 1rem'};
    }
    svg {
      width: 1.5rem;
    }
  `,
  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      opacity: 0.3;
      pointer-events: none;
      transition: none;
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, category, fullWidth, otherSide, hasIcon, disabled }) => css`
    align-items: center;
    border-radius: 1rem;
    cursor: pointer;
    display: inline-flex;
    justify-content: center;
    padding: 0 1.6rem;
    transition: background-color 250ms ease-in-out,
      border-color 250ms ease-in-out, box-shadow 250ms ease-in-out;
    ${customMedia.greaterThan('tabletUp')`
      justify-content: start;
    `}
    span {
      transition: color 250ms ease-in-out;
    }
    svg {
      transition: fill 250ms ease-in-out;
    }

    ${!!size && wrapperModifiers[size](theme)};
    ${!!category && wrapperModifiers[category](theme)};
    ${!!fullWidth && wrapperModifiers.fullWidth()};
    ${!!hasIcon && wrapperModifiers.withIcon(otherSide!)};
    ${disabled && wrapperModifiers.disabled()};
  `}
`
