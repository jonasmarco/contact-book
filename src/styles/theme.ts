export default {
  grid: {
    container: '120rem',
    defaultPadding: '1rem',
    gutter: '1.5rem'
  },
  border: {
    radius: '1rem',
    rounded: '5rem'
  },
  fonts: {
    sizes: {
      small: '1.2rem',
      normal: '1.8rem',
      large: '2.2rem',
      extraLarge: '2.8rem',
      bigger: '4.2rem'
    },
    weights: {
      slim: 300,
      normal: 400,
      bold: 600,
      extraBold: 700
    }
  },
  colors: {
    black: '#02020A',
    white: '#FCFCFC',
    red: '#FF6565',
    blue: '#4FB6FF',
    darkBlue: '#273549',
    gray: '#596E7D',
    yellow: '#F5A623',
    green: '#33CC33',

    primary: '#4FB6FF',
    secondary: '#596E7D',
    tertiary: '#30425C'
  },
  layout: {
    mobile: {
      gap: '2rem'
    },
    tablet: {
      gap: '6rem'
    },
    notebook: {
      gap: '10rem'
    },
    gapForm: '1rem'
  },
  text: {
    letterSpacing: '0.135rem',
    lineHeight: '2.8rem'
  },
  margin: {
    normal: '1.8rem',
    textMargin: '1.2rem'
  },
  layers: {
    base: 0,
    behindMenu: 5,
    menu: 10,
    modal: 100
  }
} as const
