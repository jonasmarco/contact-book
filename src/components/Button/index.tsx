import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'

import * as S from './styles'

type ButtonProps =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type Props = {
  children: React.ReactNode
  size?: 'normal' | 'large'
  category?: 'primary' | 'secondary' | 'danger' | 'success' | 'warning'
  fullWidth?: boolean
  icon?: React.ReactNode
  otherSide?: boolean
  as?: React.ElementType
} & ButtonProps

const Button = ({
  children,
  size = 'normal',
  category = 'primary',
  fullWidth = false,
  icon,
  otherSide = false,
  ...props
}: Props): JSX.Element => {
  return (
    <S.Wrapper
      size={size}
      category={category}
      fullWidth={fullWidth}
      hasIcon={!!icon}
      otherSide={otherSide}
      {...props}
    >
      {icon && otherSide && icon}
      {!!children && <span>{children}</span>}
      {icon && !otherSide && icon}
    </S.Wrapper>
  )
}

export default Button
