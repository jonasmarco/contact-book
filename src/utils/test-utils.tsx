import { ReactElement } from 'react'
import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/theme'

const customRender = (ui: ReactElement) =>
  render(
    <Router>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Router>
  )

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
