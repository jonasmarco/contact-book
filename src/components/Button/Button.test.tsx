import { render, screen, fireEvent } from '../../utils/test-utils'
import { Peace } from '@styled-icons/bootstrap'

import theme from '@/styles/theme'
import Button from '.'

describe('Button Component', () => {
  // (renderiza sem erros):
  // ensures that the Button is rendered without errors
  test('is rendered without error', () => {
    render(<Button>Click Me</Button>)
  })

  // (renderiza com o texto correto):
  // ensures that the Button is rendered with the correct text that was passed in as a child component
  it('is rendered with the correct text', () => {
    const { getByText } = render(<Button>Click Me</Button>)
    expect(getByText('Click Me')).toBeInTheDocument()
  })

  // (renderiza com o tamanho correto):
  // ensures that the Button is rendered with the correct size, based on the prop that was passed in.
  it('is rendered with the correct size', () => {
    render(<Button size="large">Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveStyle(
      'min-height: 5rem'
    )
  })

  // (renderiza com a categoria correta (secondary)):
  // ensures that the Button is rendered with the correct category, based on the prop that was passed in
  it('is rendered with the correct category', () => {
    render(<Button category="secondary">Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveStyle(`
      background-color: ${theme.colors.secondary};
      border: 1px solid ${theme.colors.secondary};
      color: ${theme.colors.white};
    `)
  })

  // (renderiza com a categoria correta (success)):
  // ensures that the Button is rendered with the correct category, based on the prop that was passed in
  it('is rendered with the correct category', () => {
    render(<Button category="success">Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveStyle(`
      background-color: ${theme.colors.green};
      border: 1px solid ${theme.colors.green};
      color: ${theme.colors.white};
    `)
  })

  // (renderiza com a categoria correta (warning)):
  // ensures that the Button is rendered with the correct category, based on the prop that was passed in
  it('is rendered with the correct category', () => {
    render(<Button category="warning">Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveStyle(`
      background-color: ${theme.colors.yellow};
      border: 1px solid ${theme.colors.yellow};
      color: ${theme.colors.white};
    `)
  })

  // (renderiza com a largura total):
  // ensures that the Button is rendered with the correct width
  it('is rendered with the correct width', () => {
    render(<Button fullWidth>Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveStyle(
      'width: 100%'
    )
  })

  // (renderiza com um ícone)
  // ensures that the Button is rendered with an icon
  it('is rendered with an icon', () => {
    render(<Button icon={<Peace data-testid="icon" />}>Click Me</Button>)
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  // (renderiza com um ícone no lado esquerdo):
  // ensure that the Button is rendered with an icon on the left
  it('is rendered with the icon on the opposite side when otherSide prop is true', () => {
    render(
      <Button icon={<Peace data-testid="icon" />} otherSide>
        Click Me
      </Button>
    )
    const button = screen.getByRole('button', { name: /Click Me/i })
    const icon = screen.getByTestId('icon')
    expect(button).toContainElement(icon)
    expect(button.firstChild).toBe(icon)
  })

  // (renderiza como desabilitado):
  // ensures that the Button is rendered as disabled when the disabled prop is passed in
  it('is rendered as disabled', () => {
    render(<Button disabled>Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveStyle(`
      cursor: not-allowed;
      opacity: 0.3;
      pointer-events: none;
    `)
  })

  // (renderiza como um link):
  // ensures that the Button is rendered as a link
  it('is rendered as a link', () => {
    render(
      <Button as="a" href="/link">
        Click Me
      </Button>
    )
    expect(screen.getByRole('link', { name: /Click Me/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })

  // (chama a função onClick quando clicado):
  // ensures that the onClick function is called when the Button is clicked
  it('calls onClick function when clicked', () => {
    const handleClick = jest.fn()
    const { getByText } = render(
      <Button onClick={handleClick}>Click Me</Button>
    )
    fireEvent.click(getByText('Click Me'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  // (renderiza com um tipo diferente de botão):
  // ensure that the Button is rendered as a different type of Button, based on the prop that was passed
  it('is rendered with a different type of button', () => {
    render(<Button type="submit">Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveAttribute(
      'type',
      'submit'
    )
  })

  // (renderiza com uma classe passada por className):
  // ensure that the Button is rendered with the class passed by className
  it('applies custom CSS class when className prop is provided', () => {
    render(<Button className="custom-button">Click Me</Button>)
    expect(screen.getByRole('button', { name: /Click Me/i })).toHaveClass(
      'custom-button'
    )
  })

  // (testa a propagação de eventos quando o botão contém elementos filhos interativos):
  // Este teste garante que quando há elementos filhos interativos no botão, os eventos de clique são propagados corretamente.
  // Neste caso, quando o elemento filho é clicado, o manipulador de cliques definido para o elemento filho é chamado exatamente uma vez.
  // ensures that if the Button has an interactive child element, the click event is correctly propagated
  it('tests event propagation when the button contains interactive child elements', () => {
    const handleChildClick = jest.fn()
    const { getByText } = render(
      <Button onClick={() => {}}>
        <span onClick={handleChildClick}>Child Element</span>
      </Button>
    )
    fireEvent.click(getByText('Child Element'))
    expect(handleChildClick).toHaveBeenCalledTimes(1)
  })

  // (snapshot)
  it('matches the snapshot in different configurations', () => {
    const { container } = render(
      <div>
        <Button>Default</Button>
        <Button size="large">Large</Button>
        <Button category="secondary">Secondary</Button>
        <Button fullWidth>Full Width</Button>
        <Button icon={<Peace />}>{}</Button>
        <Button disabled>Disabled</Button>
        <Button as="a" href="/link">
          Link
        </Button>
      </div>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
