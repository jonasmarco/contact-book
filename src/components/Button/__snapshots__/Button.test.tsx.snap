// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches the snapshot in different configurations 1`] = `
.c5 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 1.6rem;
  -webkit-transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  font-size: 1.8rem;
  min-height: 4rem;
  background: #4FB6FF;
  border: 1px solid #4FB6FF;
  color: #FCFCFC;
  position: relative;
  z-index: 5;
}

.c0 span {
  -webkit-transition: color 250ms ease-in-out;
  transition: color 250ms ease-in-out;
}

.c0 svg {
  -webkit-transition: fill 250ms ease-in-out;
  transition: fill 250ms ease-in-out;
}

.c0 span {
  line-height: 2.8rem;
}

.c0:not(:disabled):hover,
.c0:not(:disabled):focus {
  background-color: #596E7D;
  border-color: #596E7D;
  box-shadow: 0 15px 20px rgba(0,0,0,0.16),0 5px 8px rgba(0,0,0,0.08);
}

.c0:not(:disabled):hover span,
.c0:not(:disabled):focus span {
  color: #FCFCFC;
}

.c0:not(:disabled):hover svg,
.c0:not(:disabled):focus svg {
  fill: #FCFCFC;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 1.6rem;
  -webkit-transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  font-size: 2.2rem;
  min-height: 5rem;
  background: #4FB6FF;
  border: 1px solid #4FB6FF;
  color: #FCFCFC;
  position: relative;
  z-index: 5;
}

.c1 span {
  -webkit-transition: color 250ms ease-in-out;
  transition: color 250ms ease-in-out;
}

.c1 svg {
  -webkit-transition: fill 250ms ease-in-out;
  transition: fill 250ms ease-in-out;
}

.c1 span {
  line-height: 2.8rem;
}

.c1:not(:disabled):hover,
.c1:not(:disabled):focus {
  background-color: #596E7D;
  border-color: #596E7D;
  box-shadow: 0 15px 20px rgba(0,0,0,0.16),0 5px 8px rgba(0,0,0,0.08);
}

.c1:not(:disabled):hover span,
.c1:not(:disabled):focus span {
  color: #FCFCFC;
}

.c1:not(:disabled):hover svg,
.c1:not(:disabled):focus svg {
  fill: #FCFCFC;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 1.6rem;
  -webkit-transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  font-size: 1.8rem;
  min-height: 4rem;
  background: #596E7D;
  border: 1px solid #596E7D;
  color: #FCFCFC;
  position: relative;
  z-index: 5;
}

.c2 span {
  -webkit-transition: color 250ms ease-in-out;
  transition: color 250ms ease-in-out;
}

.c2 svg {
  -webkit-transition: fill 250ms ease-in-out;
  transition: fill 250ms ease-in-out;
}

.c2 span {
  line-height: 2.8rem;
}

.c2:not(:disabled):hover,
.c2:not(:disabled):focus {
  background-color: #4FB6FF;
  border-color: #4FB6FF;
  box-shadow: 0 15px 20px rgba(0,0,0,0.16),0 5px 8px rgba(0,0,0,0.08);
}

.c2:not(:disabled):hover span,
.c2:not(:disabled):focus span {
  color: #FCFCFC;
}

.c2:not(:disabled):hover svg,
.c2:not(:disabled):focus svg {
  fill: #FCFCFC;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 1.6rem;
  -webkit-transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  font-size: 1.8rem;
  min-height: 4rem;
  background: #4FB6FF;
  border: 1px solid #4FB6FF;
  color: #FCFCFC;
  position: relative;
  z-index: 5;
  width: 100%;
}

.c3 span {
  -webkit-transition: color 250ms ease-in-out;
  transition: color 250ms ease-in-out;
}

.c3 svg {
  -webkit-transition: fill 250ms ease-in-out;
  transition: fill 250ms ease-in-out;
}

.c3 span {
  line-height: 2.8rem;
}

.c3:not(:disabled):hover,
.c3:not(:disabled):focus {
  background-color: #596E7D;
  border-color: #596E7D;
  box-shadow: 0 15px 20px rgba(0,0,0,0.16),0 5px 8px rgba(0,0,0,0.08);
}

.c3:not(:disabled):hover span,
.c3:not(:disabled):focus span {
  color: #FCFCFC;
}

.c3:not(:disabled):hover svg,
.c3:not(:disabled):focus svg {
  fill: #FCFCFC;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 1.6rem;
  -webkit-transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  font-size: 1.8rem;
  min-height: 4rem;
  background: #4FB6FF;
  border: 1px solid #4FB6FF;
  color: #FCFCFC;
  position: relative;
  z-index: 5;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 span {
  -webkit-transition: color 250ms ease-in-out;
  transition: color 250ms ease-in-out;
}

.c4 svg {
  -webkit-transition: fill 250ms ease-in-out;
  transition: fill 250ms ease-in-out;
}

.c4 span {
  line-height: 2.8rem;
}

.c4:not(:disabled):hover,
.c4:not(:disabled):focus {
  background-color: #596E7D;
  border-color: #596E7D;
  box-shadow: 0 15px 20px rgba(0,0,0,0.16),0 5px 8px rgba(0,0,0,0.08);
}

.c4:not(:disabled):hover span,
.c4:not(:disabled):focus span {
  color: #FCFCFC;
}

.c4:not(:disabled):hover svg,
.c4:not(:disabled):focus svg {
  fill: #FCFCFC;
}

.c4 span {
  margin: 0 1rem 0 0;
}

.c4 svg {
  width: 1.5rem;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 1rem;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 1.6rem;
  -webkit-transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  transition: background-color 250ms ease-in-out, border-color 250ms ease-in-out,box-shadow 250ms ease-in-out;
  font-size: 1.8rem;
  min-height: 4rem;
  background: #4FB6FF;
  border: 1px solid #4FB6FF;
  color: #FCFCFC;
  position: relative;
  z-index: 5;
}

.c6 span {
  -webkit-transition: color 250ms ease-in-out;
  transition: color 250ms ease-in-out;
}

.c6 svg {
  -webkit-transition: fill 250ms ease-in-out;
  transition: fill 250ms ease-in-out;
}

.c6 span {
  line-height: 2.8rem;
}

.c6:not(:disabled):hover,
.c6:not(:disabled):focus {
  background-color: #596E7D;
  border-color: #596E7D;
  box-shadow: 0 15px 20px rgba(0,0,0,0.16),0 5px 8px rgba(0,0,0,0.08);
}

.c6:not(:disabled):hover span,
.c6:not(:disabled):focus span {
  color: #FCFCFC;
}

.c6:not(:disabled):hover svg,
.c6:not(:disabled):focus svg {
  fill: #FCFCFC;
}

.c6:disabled {
  cursor: not-allowed;
  opacity: 0.3;
  pointer-events: none;
  -webkit-transition: none;
  transition: none;
}

@media (min-width:769px) {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

@media (min-width:769px) {
  .c1 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

@media (min-width:769px) {
  .c2 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

@media (min-width:769px) {
  .c3 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

@media (min-width:769px) {
  .c4 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

@media (min-width:769px) {
  .c6 {
    -webkit-box-pack: start;
    -webkit-justify-content: start;
    -ms-flex-pack: start;
    justify-content: start;
  }
}

<div>
  <button
    class="c0"
    type="button"
  >
    <span>
      Default
    </span>
  </button>
  <button
    class="c1"
    type="button"
  >
    <span>
      Large
    </span>
  </button>
  <button
    class="c2"
    type="button"
  >
    <span>
      Secondary
    </span>
  </button>
  <button
    class="c3"
    type="button"
  >
    <span>
      Full Width
    </span>
  </button>
  <button
    class="c4"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c5"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 16 16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7.5 1.018a7 7 0 0 0-4.79 11.566L7.5 7.793V1.018zm1 0v6.775l4.79 4.79A7 7 0 0 0 8.5 1.018zm4.084 12.273L8.5 9.207v5.775a6.97 6.97 0 0 0 4.084-1.691zM7.5 14.982V9.207l-4.084 4.084A6.97 6.97 0 0 0 7.5 14.982zM0 8a8 8 0 1 1 16 0A8 8 0 0 1 0 8z"
      />
    </svg>
  </button>
  <button
    class="c6"
    disabled=""
    type="button"
  >
    <span>
      Disabled
    </span>
  </button>
  <a
    class="c0"
    href="/link"
    type="button"
  >
    <span>
      Link
    </span>
  </a>
</div>
`;
